udf/udf.proto

```diff
@@ -0,0 +1,183 @@
+// *Request messages are sent to the process from Kapacitor.
+// *Response messages are sent to Kapacitor from the process.
+//
+// While there is an obvious request/response structure for communicating,
+// there is a loose coupling between request and response.
+// Meaning that ordering or synchronizing STDIN and STDOUT in anyway
+// is not necessary.
+// For example if Kapacitor requests a state snapshot and the
+// process is in the middle of writing a previous response or
+// data points those can continue. Eventually Kapacitor will receive
+// the state snapshot response and act accordingly.
+//
+// A KeepAliveRequest/KeepAliveResponse system is used to ensure that 
+// the process is responsive. Every time that a KeepAliveRequest is sent
+// a KeepAliveResponse must be returned within a timeout. If too many heartbeats
+// are missed than the process is considered dead and will be terminated/restarted.
+
+
+// Request that the process initialize itself with the provided options.
+message InitializeRequest {
+    repeated Option options = 1;
+}
+message Option {
+    string name = 1;
+    oneof value {
+        string strValue    = 2;
+        int64  intValue    = 3;
+        double doubleValue = 4;
+    }
+}
+
+// Respond to Kapacitor whether initialization was successful.
+message InitializeResponse {
+    bool success = 1;
+    string error = 2;
+}
+
+// Request that the process provide a snapshot of its state.
+message StateRequest {
+}
+
+// Respond to Kapacitor with a versioned and serialized
+// snapshot of the running state.
+message StateResponse {
+    int32 version  = 1;
+    bytes snapshot = 2;
+}
+
+// Request that the process restore its state from a snapshot.
+message RestoreRequest {
+    int32 version  = 1;
+    bytes snapshot = 2;
+}
+
+// Respond with success or failure to RestoreRequest
+message RestoreResponse {
+    int32 version = 1;
+    bool success  = 2;
+    string error  = 3;
+}
+
+
+// Request that the process respond with a KeepAlive
+// to verify it is responding.
+message KeepAliveRequest {
+    int64 counter = 1;
+}
+
+// Respond to KeepAliveRequest
+message KeepAliveResponse {
+    int64 counter = 1;
+}
+
+// Request to stop the process
+message StopRequest {
+}
+
+
+// Sent from process to Kapacitor indicating an error has occured.
+message ErrorResponse {
+    string error = 1;
+}
+
+//------------------------------------------------------
+// Data flow messages
+//
+// Sent and received by both the process and Kapacitor
+
+
+// Indicates the beginning of a batch.
+// All subsequent points should be considered
+// part of the batch until EndBatch arrives.
+message BeginBatch {
+}
+
+// Message containing information about a single data point.
+// Can be sent on it's own or bookended by BeginBatch and EndBatch messages.
+message Point {
+    int64 Time                       = 1;
+    string Name                      = 2;
+    string Database                  = 3;
+    string RetentionPolicy           = 4;
```

Comment:
I like that idea, proto is sparse encoding so if the field is not set it doesn't take any space on the wire but maybe grouping things into messages would make that easier to manage. As opposed to having to check each field explicitly.